client/asset/dcr/dcr.go

```diff
@@ -5119,28 +5126,142 @@ func (dcr *ExchangeWallet) SetVSP(url string) error {
-		return nil, errors.New("not connected, login first")
+		return errors.New("not connected, login first")
 	}
 	// I think we need to set this, otherwise we probably end up with default
 	// of DefaultRelayFeePerKb = 1e4 => 10 atoms/byte.
 	feePerKB := dcrutil.Amount(dcr.feeRateWithFallback(feeSuggestion) * 1000)
 	if err := dcr.wallet.SetTxFee(dcr.ctx, feePerKB); err != nil {
-		return nil, fmt.Errorf("error setting wallet tx fee: %w", err)
+		return fmt.Errorf("error setting wallet tx fee: %w", err)
+	}
+
+	remain := dcr.ticketBuyer.remaining.Add(int32(n))
+	dcr.emit.Data(ticketDataRoute, &TicketPurchaseUpdate{Remaining: uint32(remain)})
+	go dcr.runTicketBuyer()
+
+	return nil
+}
+
+const ticketDataRoute = "ticketPurchaseUpdate"
+
+// TicketPurchaseUpdate is an update from the asynchronous ticket purchasing
+// loop.
+type TicketPurchaseUpdate struct {
+	Err       string          `json:"err,omitempty"`
+	Remaining uint32          `json:"remaining"`
+	Tickets   []*asset.Ticket `json:"tickets"`
+}
+
+// runTicketBuyer attempts to buy requested tickets. Because of a dcrwallet bug,
+// its possible that (Wallet).PurchaseTickets will purchase fewer tickets than
+// requested, without error. To work around this bug, we add requested tickets
+// to ExchangeWallet.ticketBuyer.remaining, and re-run runTicketBuyer every
+// block.
+func (dcr *ExchangeWallet) runTicketBuyer() {
+	tb := &dcr.ticketBuyer
+	if !tb.running.CompareAndSwap(false, true) {
+		// already running
+		return
+	}
+	defer tb.running.Store(false)
+	var ok bool
+	defer func() {
+		if !ok {
+			tb.remaining.Store(0)
+		}
+	}()
+
+	if tb.unconfirmedTickets == nil {
+		tb.unconfirmedTickets = make(map[chainhash.Hash]struct{})
+	}
+
+	remain := tb.remaining.Load()
+	if remain < 1 {
+		return
+	}
+
+	var unconf int
+	for txHash := range tb.unconfirmedTickets {
+		tx, err := dcr.wallet.GetTransaction(dcr.ctx, &txHash)
+		if err != nil {
+			dcr.log.Errorf("GetTransaction error ticket tx %s: %v", txHash, err)
+			dcr.emit.Data(ticketDataRoute, &TicketPurchaseUpdate{Err: err.Error()})
+			return
+		}
+		if tx.Confirmations > 0 {
+			delete(tb.unconfirmedTickets, txHash)
+		} else {
+			unconf++
+		}
+	}
+	if unconf > 0 {
+		ok = true
+		dcr.log.Tracef("Skipping ticket purchase attempt because there are still %d unconfirmed tickets", unconf)
+	}
+
+	dcr.log.Tracef("Attempting to purchase %d tickets", remain)
+
+	bal, err := dcr.Balance()
+	if err != nil {
+		dcr.log.Errorf("GetBalance error: %v", err)
+		dcr.emit.Data(ticketDataRoute, &TicketPurchaseUpdate{Err: err.Error()})
+		return
+	}
+	sinfo, err := dcr.wallet.StakeInfo(dcr.ctx)
+	if err != nil {
+		dcr.log.Errorf("StakeInfo error: %v", err)
+		dcr.emit.Data(ticketDataRoute, &TicketPurchaseUpdate{Err: err.Error()})
+		return
 	}
+	if dcrutil.Amount(bal.Available) < sinfo.Sdiff*dcrutil.Amount(remain) {
+		dcr.log.Errorf("Insufficient balance %s to purches %d ticket at price %s: %v", dcrutil.Amount(bal.Available), remain, sinfo.Sdiff)
```

Comment:
```
2023-11-17 05:05:30.709 [ERR] CORE[dcr]: Insufficient balance 0 DCR to purches 990 ticket at price 0.0002 DCR: %!v(MISSING)
```

The last %v doesn't have an error.

Also, this was waiting for confs so tried a second time. I think returning at line 5205 prevents this.
```
2023-11-17 05:05:30.709 [TRC] CORE[dcr]: Attempting to purchase 990 tickets
2023-11-17 05:05:30.709 [TRC] CORE: Processing tip change for dcr
2023-11-17 05:05:30.709 [ERR] CORE[dcr]: Insufficient balance 0 DCR to purches 990 ticket at price 0.0002 DCR: %!v(MISSING)
2023-11-17 05:05:34.041 [DBG] CORE[btc]: tip change: 454 (33cb50bc8c3379bd6443e4cbef3611b1a3c665d6506dbb6246632d147dac929c) => 455 (0b42f1dd56b0a8a5e76bb68e34bb9c984a44f9c6d15b4aebb716a05a7a3bb6f0)
2023-11-17 05:05:34.041 [TRC] CORE: Processing tip change for btc
2023-11-17 05:05:34.041 [TRC] CORE[btc][SPV]: Bals: spendable = 0 BTC (0 BTC trusted, 0 BTC untrusted, 0 BTC assumed locked), immature = 0 BTC
2023-11-17 05:05:46.281 [DBG] CORE[dcr]: tip change: 651 (31900f90988e23659ec301bcf5f39bc98eabfea389babff7dfe79b91412ce0ad) => 652 (b97d6ba3f36a6e589d33e84f4c09eb9f2254add1bbbbbfbc1be334bdfb5d0704)
2023-11-17 05:05:46.281 [TRC] CORE: Processing tip change for dcr
2023-11-17 05:05:46.281 [TRC] CORE[dcr]: Attempting to purchase 990 tickets
2023-11-17 05:05:46.281 [ERR] CORE[dcr]: Insufficient balance 0 DCR to purches 990 ticket at price 0.0002 DCR: %!v(MISSING)
```